File: Easy/0977_squares-of-a-sorted-array.py
"""
977. Squares of a Sorted Array
https://leetcode.com/problems/squares-of-a-sorted-array/

@Easy
@Algorithm

Given an integer array nums sorted in non-decreasing order, return an array of the squares of each number sorted in non-decreasing order.
"""
from typing import List


# v1 58.01%


class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        return sorted([n**2 for n in nums])


# v2


class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        res = list()
        left, right = 0, len(nums) - 1
        for _ in range(len(nums)):
            if abs(nums[left]) <= abs(nums[right]):
                index = right
                right -= 1
            else:
                index = left
                left += 1
            res.insert(0, nums[index]**2)
        return res


sol = Solution()
print(f"{sol.sortedSquares(nums = [-4,-1,0,3,10]) = }")
print(f"{sol.sortedSquares(nums = [-4, -2, 0]) = }")

File: Easy/0977_squares-of-a-sorted-array.cpp
/*
977. Squares of a Sorted Array
https://leetcode.com/problems/squares-of-a-sorted-array/

@Easy
@Algorithm

Given an integer array nums sorted in non-decreasing order, return an array of the squares of each number sorted in non-decreasing order.
*/

#include <iostream>
#include <vector>
using namespace std;

class Solution {
   public:
    vector<int> sortedSquares(vector<int>& nums) {
        vector<int> res(nums.size());
        int l = 0, r = nums.size() - 1;
        for (int i = nums.size() - 1; i >= 0; i--) {
            if (abs(nums[l]) > abs(nums[r])) {
                res[i] = nums[l] * nums[l];
                l++;
            } else {
                res[i] = nums[r] * nums[r];
                r--;
            }
        }
        return res;
    }
    void sortedSquares(int nums[], int size) {
        vector<int> res(size);
        int l = 0, r = size - 1;
        for (int i = size - 1; i >= 0; i--) {
            if (abs(nums[l]) > abs(nums[r])) {
                res[i] = nums[l] * nums[l];
                l++;
            } else {
                res[i] = nums[r] * nums[r];
                r--;
            }
        }
        copy(res.begin(), res.end(), nums);
    }
};

int main() {
    Solution sol;

    vector<int> nums{10, 2, 30};
    nums = sol.sortedSquares(nums);

    for (int x : nums) {
        cout << x << " ";
    }
    cout << "\n";

    int nums_array[] = {10, 2, 30};
    int size = (sizeof(nums_array) / sizeof(nums_array[0]));
    sol.sortedSquares(nums_array, size);

    for (int x : nums_array) {
        cout << x << " ";
    }
    cout << "\n";
    return 0;
}
